sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -366,22 +367,34 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
             // ((c || ...) && (d || ...)) || a || b
             (common :+ And(ldiff.reduce(Or), rdiff.reduce(Or))).reduce(Or)
           }
+        } else {
+          // No common factors from disjunctive predicates, reduce common factor from conjunction
+          val all = splitConjunctivePredicates(left) ++ splitConjunctivePredicates(right)
+          val distinct = ExpressionSet(all)
+          if (all.size == distinct.size) {
+            // No common factors, return the original predicate
+            and
+          } else {
+            // (((a && b) && a && (a && c))) => a && b && c
+            distinct.reduce(And)
```

Comment:
one concern is, if there are many conjunctive predicates, here we may build a left-deep tree while the original one is balanced. Can we add a util function in `PredicateHelper` to build a balanced And/Or tree?

cc @gengliangwang

Reply:
@cloud-fan yes that would be helpful, especially in filter pushdown of data sources.

Reply:
thanks @cloud-fan, good observation. I have added required Utils to address this.